flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherResourceCleanupTest.java

```diff
@@ -651,6 +654,91 @@ public void testFailingJobManagerRunnerCleanup() throws Exception {
         awaitStatus(dispatcherGateway, jobId, JobStatus.RUNNING);
     }
 
+    @Test
+    public void testArchiveSuccessfullyWithinTimeout() throws Exception {
+
+        final long archiveTimeout = 1000L;
+        final CompletableFuture<Void> archiveFuture = new CompletableFuture<>();
+        final TestingHistoryServerArchivist archivist =
+                new TestingHistoryServerArchivist(archiveFuture);
+
+        final TestingJobManagerRunner testingJobManagerRunner =
+                startDispatcherAndSubmitJob(archiveTimeout, archivist);
+        finishJob(testingJobManagerRunner);
+
+        // Before the archiving is finished, the cleanup is not finished and the job is not
+        // terminated.
+        assertThatNoCleanupWasTriggered();
+        final CompletableFuture<Void> jobTerminationFuture =
+                dispatcher.getJobTerminationFuture(jobId, Time.milliseconds(1000L));
+        assertFalse(jobTerminationFuture.isDone());
+
+        archiveFuture.complete(null);
+
+        // Once the archive is finished, the cleanup is finished and the job is terminated.
+        assertGlobalCleanupTriggered(jobId);
+        jobTerminationFuture.join();
+
+        assertTrue(archivist.isArchived());
+    }
+
+    @Test
+    public void testArchiveFailedWhenTimeoutExceeded() throws Exception {
+
+        final long archiveTimeout = 10L;
+        final CompletableFuture<Void> archiveFuture = new CompletableFuture<>();
+        final TestingHistoryServerArchivist archivist =
+                new TestingHistoryServerArchivist(archiveFuture);
+
+        final TestingJobManagerRunner testingJobManagerRunner =
+                startDispatcherAndSubmitJob(archiveTimeout, archivist);
+
+        finishJob(testingJobManagerRunner);
+
+        // Once the archiving is timed out, the cleanup will be finished and the job will be
+        // terminated.
+        assertGlobalCleanupTriggered(jobId);
+        dispatcher.getJobTerminationFuture(jobId, Time.milliseconds(1000L)).join();
+
+        assertFalse(archivist.isArchived());
+    }
+
+    @Test
+    public void testNotArchiveSuspendedJob() throws Exception {
+
+        final CompletableFuture<Void> archiveFuture = new CompletableFuture<>();
+        final TestingHistoryServerArchivist archivist =
+                new TestingHistoryServerArchivist(archiveFuture);
+
+        final TestingJobManagerRunner testingJobManagerRunner =
+                startDispatcherAndSubmitJob(1000L, archivist);
+
+        suspendJob(testingJobManagerRunner);
+
+        assertLocalCleanupTriggered(jobId);
+        dispatcher.getJobTerminationFuture(jobId, Time.milliseconds(1000L)).join();
+
+        assertFalse(archivist.isArchived());
+    }
+
+    private TestingJobManagerRunner startDispatcherAndSubmitJob(
```

Comment:
I'm not sure whether we actually need that method. It does everything `startDispatcherAndSubmitJob` already does (i.e. instantiating a `TestingJobMasterServiceLeadershipRunnerFactory`, creating the `Dispatcher` and starting it. The `CLUSTER_SERVICES_SHUTDOWN_TIMEOUT` only has to be set once, i.e. we can pass in the `TestingDispatcher.Builder` instance and set the `HistoryServerArchivist` in the corresponding test (which is essentially only the callback). WDYT?